substrate/network/src/consensus.rs

```diff
@@ -0,0 +1,203 @@
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.?
+
+//! Consensus related bits of the network service.
+
+use std::collections::HashMap;
+use multiqueue;
+use futures::sync::oneshot;
+use io::SyncIo;
+use protocol::Protocol;
+use network::PeerId;
+use primitives::Hash;
+use message::{self, Message};
+use runtime_support::Hashable;
+
+//TODO: The queue is preallocated in multiqueue. Make it unbounded
+const QUEUE_SIZE: u64 = 1 << 16;
+
+struct CandidateRequest {
+	id: message::RequestId,
+	completion: oneshot::Sender<Vec<u8>>,
+}
+
+struct PeerConsensus {
+	candidate_fetch: Option<CandidateRequest>,
+	candidate_available: Option<Hash>,
+}
+
+/// Consensus network protocol handler. Manages statements and candidate requests.
+pub struct Consensus {
+	peers: HashMap<PeerId, PeerConsensus>,
+	our_candidate: Option<(Hash, Vec<u8>)>,
+	statement_sink: multiqueue::BroadcastFutSender<message::Statement>,
+	statement_stream: multiqueue::BroadcastFutReceiver<message::Statement>,
+	bft_message_sink: multiqueue::BroadcastFutSender<message::BftMessage>,
+	bft_message_stream: multiqueue::BroadcastFutReceiver<message::BftMessage>,
+}
+
+impl Consensus {
+	/// Create a new instance.
+	pub fn new() -> Consensus {
+		let (statement_sink, statement_stream) = multiqueue::broadcast_fut_queue(QUEUE_SIZE);
+		let (bft_sink, bft_stream) = multiqueue::broadcast_fut_queue(QUEUE_SIZE);
+		Consensus {
+			peers: HashMap::new(),
+			our_candidate: None,
+			statement_sink: statement_sink,
+			statement_stream: statement_stream,
+			bft_message_sink: bft_sink,
+			bft_message_stream: bft_stream,
+		}
+	}
+
+	/// Closes all notification streams.
+	pub fn restart(&mut self) {
+		let (statement_sink, statement_stream) = multiqueue::broadcast_fut_queue(QUEUE_SIZE);
+		let (bft_sink, bft_stream) = multiqueue::broadcast_fut_queue(QUEUE_SIZE);
+		self.statement_sink = statement_sink;
+		self.statement_stream = statement_stream;
+		self.bft_message_sink = bft_sink;
+		self.bft_message_stream = bft_stream;
+	}
+
+	/// Handle new connected peer.
+	pub fn new_peer(&mut self, _io: &mut SyncIo, _protocol: &Protocol, peer_id: PeerId, roles: &[message::Role]) {
+		if roles.iter().any(|r| *r == message::Role::Validator) {
+			trace!(target:"sync", "Registering validator {}", peer_id);
+			self.peers.insert(peer_id, PeerConsensus {
+				candidate_fetch: None,
+				candidate_available: None,
+			});
+		}
+	}
+
+	pub fn on_statement(&mut self, peer_id: PeerId, statement: message::Statement) {
+		if let Some(ref mut peer) = self.peers.get_mut(&peer_id) {
+			// TODO: validate signature?
```

Comment:
also TODO (but beyond this PR) is handle only current parent-hash statement

Reply:
This will be addressed in an upcoming PR